### ISSUE TITLE
port git-status-cache to linux, macintosh

### ISSUE THREAD

**Issue by BernMcCarty** [cmarcusreid/git-status-cache#23]
Might this project port to linux and the mac?

git and much around it is portable: dotnet core, libgit2sharp, powershell core, etc. I just set up posh-git from its master branch on linux and it seems very slow in my VM environment. It looks like a port of [GitStatusCachePoshClient](https://github.com/cmarcusreid/git-status-cache-posh-client) would probably make a big difference.

**Comment by dahlbyk** [comment]
@BernMcCarty can you enable `$GitPromptSettings.Debug = $true` and share what you're seeing for timings?

**Comment by BernMcCarty** [comment]
Hopefully this is what you were looking for:

VERBOSE: 00067:Getting status
VERBOSE: 00369:Parsing status
VERBOSE: 00371:Status: ## master...origin/master
VERBOSE: 00375:Status:  M BuildInf/Install-NvcPoshModules.ps1
VERBOSE: 00379:Status:  M BuildInf/LoadPoshGit.ps1
VERBOSE: 00390:Building status object
VERBOSE: 00484:Finished

**Comment by BernMcCarty** [comment]
@dahlbyk sadly no, not at the moment.  Right now I am only using this version of posh-git on an Ubuntu guest of a Windows 7 virtualbox host.  The host machine is pretty good but the .vdi is on over on an unimpressive spinny disk, so performance is wanting generally.  I started to try to bring GitStatusCachePoshClient into the equation to try to get some better performance before I even realized that it was windows only.

**Comment by dahlbyk** [comment]
👍 thanks for clarifying. I'll defer to @cmarcusreid as to how hard it would be make this work x-plat, but at least the libgit2 portion of it should work fine cross-platform.

Down with spinning rust!

**Comment by cmarcusreid** [comment]
I don't currently have plans to implement cross-platform support, but it'd be a welcome contribution!

Most of the logic is platform agnostic. I think there are two pieces to figure out:
1.) The file sytem watcher (ext/ReadDirectoryChanges/) uses raw Windows APIs.
2.) The cache executable and PowerShell module communicate over named pipes.

**Comment by hcoona** [comment]
I'd like to share some of my knowledge.

For the platform agnostic parts, we have some stable alternatives:
1. The file system watcher: https://facebook.github.io/watchman/
2. The named pipes. https://man7.org/linux/man-pages/man7/fifo.7.html & https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/mkfifo.2.html. But I think it more reasonable to communicate with standard RESTful API instead of named pipe. It's easy to understand, standardized, easy to debug, and accessible cross-platform.

**Comment by hcoona** [comment]
I'm considering rewriting GitStatusCache with C#, but there's no time plan.

**Comment by hcoona** [comment]
I also wonder if we still need this project after git supported external monitor hook, see "Using Watchman and the sample hook script" in https://github.blog/2022-06-29-improve-git-monorepo-performance-with-a-file-system-monitor/

**Comment by dahlbyk** [comment]
> "Using Watchman and the sample hook script" in https://github.blog/2022-06-29-improve-git-monorepo-performance-with-a-file-system-monitor/

TIL! We should add a note about this to the posh-git readme.

TL;DR, upgrade to Git 2.37+ and:

```pwsh
git config --global core.fsmonitor true
git config --global core.untrackedcache true
```

(If you set `core.fsmonitor true` on a slightly older version, Git and posh-git get confused.)

**Comment by hcoona** [comment]
@dahlbyk be careful to enable it globally if you have many repos in your machine. I believe the builtin fsmonitor would launch background processes for each repo. The solution I mentioned would only launch a centralized background process.

BTW, I also think it boring to setup fsmonitor. I use [microsoft scalar](https://devblogs.microsoft.com/devops/introducing-scalar/), it has a [`scalar register` command](https://github.com/microsoft/git/blob/ffd33cfd1823d78f4159d4e3e7bf691a5aacd422/contrib/scalar/docs/getting-started.md#registering-existing-git-repos) for that. You can install it by install the latest version of [microsoft maintained git](https://github.com/microsoft/git).
